doc/sphinx/proofs/writing-proofs/inductive.rst

```diff
@@ -186,56 +195,52 @@ analysis on inductive or co-inductive objects (see :ref:`inductive-definitions`)
+   .. tacv:: destruct @term {? with @bindings } {? @as_or_and_ipat } {? eqn:@naming_intropattern } {? using @term {? with @bindings } } {? in @goal_occurrences }
+             edestruct @term {? with @bindings } {? @as_or_and_ipat } {? eqn:@naming_intropattern } {? using @term {? with @bindings } } {? in @goal_occurrences }
 
       These are the general forms of :tacn:`destruct` and :tacn:`edestruct`.
       They combine the effects of the ``with``, ``as``, ``eqn:``, ``using``,
       and ``in`` clauses.
 
-.. tacn:: case @term
-   :name: case
+.. tacn:: case @induction_clause_list
 
    The tactic :n:`case` is a more basic tactic to perform case analysis without
-   recursion. It behaves as :n:`elim @term` but using a case-analysis
-   elimination principle and not a recursive one.
+   recursion. It behaves like :n:`elim @term` but uses a case-analysis
+   elimination principle rather than a recursive one.
 
 .. tacv:: case @term with @bindings
 
    Analogous to :n:`elim @term with @bindings` above.
 
-.. tacv:: ecase @term {? with @bindings }
-   :name: ecase
+.. tacn:: ecase @induction_clause_list
 
-   In case the type of :n:`@term` has dependent premises, or dependent premises
-   whose values are not inferable from the :n:`with @bindings` clause,
+   If the type of :n:`@term` has dependent premises, or dependent premises
+   whose values are not inferrable from the :n:`with @bindings` clause,
    :n:`ecase` turns them into existential variables to be resolved later on.
 
-.. tacv:: simple destruct @ident
-   :name: simple destruct
+.. tacn:: simple destruct {| @ident | @natural }
 
-   This tactic behaves as :n:`intros until @ident; case @ident` when :n:`@ident`
+   Equivalent to :n:`intros until @ident; case @ident` when :n:`@ident`
    is a quantified variable of the goal.
 
 .. tacv:: simple destruct @natural
 
-   This tactic behaves as :n:`intros until @natural; case @ident` where :n:`@ident`
+   Behaves like :n:`intros until @natural; case @ident` where :n:`@ident`
    is the name given by :n:`intros until @natural` to the :n:`@natural` -th
    non-dependent premise of the goal.
 
-.. tacv:: case_eq @term
+.. tacn:: case_eq @one_term
 
-   The tactic :n:`case_eq` is a variant of the :n:`case` tactic that allows to
-   perform case analysis on a term without completely forgetting its original
+   The tactic :n:`case_eq` is a variant of the :n:`case` tactic that allows
+   performing case analysis on a term without completely forgetting its original
    form. This is done by generating equalities between the original form of the
    term and the outcomes of the case analysis.
 
-.. tacn:: induction @term
-   :name: induction
+.. tacn:: induction @induction_clause_list
 
-   This tactic applies to any goal. The argument :n:`@term` must be of
+   The argument :n:`@one_term` in :n:`@induction_clause_list` must be of
    inductive type and the tactic :n:`induction` generates subgoals, one for
```

Comment:
The induction hypothesis _always_ looks just like the theorem statement?  If so I maybe I should state that in a clear way.  Ex: `forall n, n + 0 = n` gives you the induction hypothesis `IHn : n + 0 = n`.

My initial thought was to look in the [Simple Inductive Types section](https://coq.inria.fr/distrib/current/refman/language/core/inductive.html#simple-inductive-types), where it talks about "elimination principles", which is barely described in the doc.  How does `nat_rec` relate to the induction hypothesis?

Reply:
>The induction hypothesis always looks just like the theorem statement?

No. It depends on the inductive type definition. `nat` is a very simple case.

>  How does `nat_rec` relate to the induction hypothesis?

It doesn't. `nat_ind` does.

Reply:
Doesn't `induction` use `nat_rec` if your goal is in `Set` instead of `Prop`?

Reply:
Indeed, it does.

Reply:
I guess the most relevant thing would be to explain how the type of `nat_ind` from `Check nat_ind.` relates to the behavior of the tactic.  At least as an example.

```
nat_ind : 
  forall P : nat -> Prop,
  P 0 ->
  (forall n : nat, P n -> P (S n)) ->
  forall n : nat, P n
```

But it seems to me there be a lot that's unsaid about the various elimination principles that inductives generate, e.g. for `nat`:

```
nat is defined
nat_rect is defined
nat_ind is defined
nat_rec is defined
nat_sind is defined
```

Reply:
Also, if `nat` is `Set`, then are the other 3 elimination principles ever (or commonly) useful for `nat`?

Reply:
The difference in induction principles is about the Sort of the goal you are proving, not about the Sort of `nat`.

`induction X using lem` is roughly the same as `generalize dependent X; apply lem; intros <whatever was generalized>`, modulo some checking about arguments to `lem`.  This should explain how the type of induction principles relates to the type of the goal and the types of the generated goals, roughly.  Suppose the goal is the type `G` and the inductive family of `X` is named `T` and that no `using` clause was passed; then `lem` defaults to `T_ind` if the Sort of `G` is `Prop`, `T_rec` when it is `Set`, `T_rect` if it is `Type`, and `T_sind` if the Sort of `G` is `SProp`.

Reply:
Also, I notice some of the doc uses the term "elimination principle", such as [here](https://coq.inria.fr/distrib/current/refman/language/core/inductive.html#simple-inductive-types).  Seems to me it would be more intuitive to refer to these as "induction principles" and add that term to the glossary.  WDYT?

Reply:
Ah, the `using` clause in `induction` and `destruct` explicitly provides the induction principle?  If so, I will document that.  Pretty sure that wasn't in the original text.

Q1: `induction` and `destruct` are both needed? IIUC the difference between them is slight.  Which is used more/is more important?  Can we give any non-trivial guidance on when to use one or the other?  (IIUC, you must use `destruct` for co-induction.  See [here](https://coq.inria.fr/distrib/current/refman/language/core/coinductive.html#coq:cmd.CoInductive).)  Perhaps `induction` should appear in the doc before `destruct` (in part because `induction` is more obviously related to "Reasoning with inductive types")?  

Q2: Is it common to have a user-defined induction principle?  Is the purpose to do things like [prefix induction](https://en.wikipedia.org/wiki/Mathematical_induction#Prefix_induction)?

Reply:
>Also, I notice some of the doc uses the term "elimination principle", such as here. Seems to me it would be more intuitive to refer to these as "induction principles" and add that term to the glossary. WDYT?

I guess elimination principle is in theory more general (also applies when there is no recursion) but uniformizing toward induction principle could make sense in the context of Coq.

> Pretty sure that wasn't in the original text.

Here: https://coq.github.io/doc/master/refman/proofs/writing-proofs/reasoning-inductives.html#coq:tacv.induction-%E2%80%A6-using-%E2%80%A6

Q1: both are equally important. You don't want to generate induction hypotheses when you don't need them. You don't want your proof term to contain a `fix` when only a `match` was necessary. Just like you don't want to use `Fixpoint` when all you needed was `Definition`. I don't get your reference to co-induction (but co-induction is anyway a special case that is much less well-supported in Coq).

>  Perhaps `induction` should appear in the doc before `destruct` (in part because `induction` is more obviously related to "Reasoning with inductive types")?

No, `destruct` should appear first. In fact, we should probably separate the section in two "Case analysis" then "Induction". Here the title of the chapter is encompassing, in particular of non-recursive inductive types (for which `induction` is useless) and of co-inductive types (for which there is the `cofix` tactic).

Q2: Hugo gave an example in another thread for double induction. I guess the answer (of how common it is) is it depends on the users.

Reply:
I changed some of the uses of "elimination principle" to "induction principle" and made the latter a defined term, but I didn't change all 47 uses of the latter to use `:term:`--I expect the term is mostly well-understood.  Though I could certainly change them all.

That led me to wonder whether the "Generation of Induction Principles with Scheme" section might be better in the "Core Language/Inductive types" chapter, which already has quite a bit material about generation induction principles.  Also that section has a bunch of commands in it whereas the "Reasoning with inductive types" section seems like it should focus on tactics.  Also maybe move the "Generation of inversion principles" section to the other chapter.  What do you think?  Unfortunately, the Inductive types section is already fairly long.

Also, the explanation for the second example (odd and even) after [here](https://coq.inria.fr/distrib/current/refman/user-extensions/proof-schemes.html#coq:cmd.Scheme) seems incomplete--when would you use the induction principles `odd_ind`/`even_ind` versus `odd_even`/`even_odd`?  The latter would have to explicit on an `induction` tactic.  And the examples of induction principles should show their type more often and perhaps say just a little about the principle is generated (AFAICT one subterm per constructor).  Perhaps some of this could be changed outside of this PR.

Reply:
Hardly a model of clarity.  "term as induction scheme" doesn't make it clear that it refers to the second "term" in the grammar.  I'll change `induction scheme` to `induction principle` everywhere.

![image](https://user-images.githubusercontent.com/1253341/121289122-582fa100-c899-11eb-8922-5068dc9d64cf.png)

Reply:
> That led me to wonder whether the "Generation of Induction Principles with Scheme" section might be better in the "Core Language/Inductive types" chapter, which already has quite a bit material about generation induction principles. Also that section has a bunch of commands in it whereas the "Reasoning with inductive types" section seems like it should focus on tactics. Also maybe move the "Generation of inversion principles" section to the other chapter. What do you think? Unfortunately, the Inductive types section is already fairly long.

This would really go against the spirit of the "Core language" chapter which was meant to be focused on the kernel language, which the `Scheme` commands have nothing to do with (they have everything to do with reasoning with inductive types). I would rather go in the other direction and move part of the explanation currently in "Core language" to here (but we can think of that later).

> Also, the explanation for the second example (odd and even) after [here](https://coq.inria.fr/distrib/current/refman/user-extensions/proof-schemes.html#coq:cmd.Scheme) seems incomplete--when would you use the induction principles `odd_ind`/`even_ind` versus `odd_even`/`even_odd`?

Currently, auto-generated induction principles for mutual inductive types are useless, so they are never used. This could be spelled out more clearly in the doc.

Reply:
I added a statement about the useless IPs.

I notice that the tree/forest examples [here](https://coq.github.io/doc/master/refman/language/core/inductive.html#mutually-defined-inductive-types) and [here](https://coq.github.io/doc/master/refman/proofs/writing-proofs/reasoning-inductives.html#coq:cmd.Combined-Scheme) are almost identical.  I'll drop the first one but I'll use `Parameter`s rather than `Axiom`s.

Reply:
On second thought, I'll hold off on removing one of the near duplicates.

Reply:
Would it make sense to enhance `Inductive` so it defines a useful induction principle for mutually inductive types?

Reply:
Yes.  I suspect most of the issue is just finding the time to write it, though it's possible there are some subtleties that need to be figured out

Reply:
The main reason why it has been hold off was not a lack of time but the perfect-is-the-enemy-of-good problem. Coq developers wanted to figure out the most generic mutual induction principles before changing what `Inductive` generates. But this strategy has turned out to be highly detrimental to the users. So today, the developers would support anyone who would integrate the mutual induction principles that `Scheme` generates in the default mechanism used by `Inductive`.